README

#=========================================
Tseries - Meteorological Timeseries Viewer
#=========================================

Norwegian Meteorological Institute (met.no)
Box 43 Blindern
0313 OSLO
NORWAY
email: diana@met.no

Contents:
1. About Tseries
2. Install
3. Running Tseries
4. Configuration files
5. File formats

#-------------------------------------------
# 1. About Tseries
#-------------------------------------------

Tseries is part of the Diana package from met.no (Norwegian Meteorological Institute) and is distributed under the GPL license. See gpl.txt for details concerning the GPL license.

Tseries is a graphical time series viewer developed for use with meteorological and oceanographic data. It uses OpenGL for the graphics and Trolltech Qt for the user interface.

Tseries supports multiple languages via the Qt linguist system. Additional languages can be added by preparing new lang/tseries_XX.ts files (using the Qt linguist program) and adding the new language in the proper place in the Makefiles. 

In addition to the interactive viewer, there is also a command line version of Tseries for batch plotting of time series diagrams (btseries).

Depends on:
Trolltech Qt
OpenGl
HDF 4.2


#-------------------------------------------
# 2. Install
#-------------------------------------------


The included libraries needed for Tseries:
get metlibs from https://svn.met.no/metlibs/
compile og install them in a proper place.

export PACKAGE_CONFIG_PATH=where_you_installed_metlibs/lib/pkgconfig

./autogen.sh
./configure
make; 
make install

#-------------------------------------------
# 3. Running Tseries
#-------------------------------------------

running Tseries with the prepared test datasets and configurations:

1. The test data sets should be installed under metnolocal/demodata/

2. Starting Tseries:
cd metnolocal/tseries
bin/tseries -s tseries.ctl_export

Select the language of preference under "Preferences | languages" in the menu.
User guide available under Help.


running btseries:
The batch version of this plotting package is called btseries and will be installed in the same place as Tseries. Try "bin/btseries" for command line help, and "bin/btseries -example" to dump a complete job control file.

#-------------------------------------------
# 4. Configuration files
#-------------------------------------------

tseries.ctl_export:
  The setup file for Tseries. 
  Change top level directory for installation with 'Topdir' variable
  Add new datafiles under section <streams> 

etc/tsDiagrams_export.def:
  Definition of diagram types with supported data
  See comments in file for syntax
  IMPORTANT: if you change the top level directory in tseries.ctl_export,
  you must also update the style paths here (see STYLEFILE=)

.tseries.conf:
  User preferences saved here.

etc/tseries.filter:
  User defined position filter saved here (see online doc)

etc/symbols.def:
  Weather symbol mapping

style/style.*
  Detailed instructions for the plotting package (in ascii). 
  Determines which components to add to a specific diagram.
  Used in etc/tsDiagrams_export.def. The structure of these files
  is currently undocumented.

#-------------------------------------------
# 5. File formats
#-------------------------------------------

Tseries supports a number of in-house (met.no) formats. 

1. The most commonly used is the HDF 4.2 based format. In the <streams> section of the setup file you will find references to files of this type with "DataType=HDF". 
The exact specification for this format is omitted here; mainly because we intend to make changes to this format in the near future. If you are still interested: use the 'vshow' program that comes with the HDF4.2 package to view the structure of one the files in the test data set.

2. Tseries also supports a simple ascii format. See the test data file metnolocal/demodata/hirlam/t2m_lqr.txt for a complete example. 
Here is a short definition:

# --------------------------------------
# Example time series data file in ascii
# --------------------------------------

#- source name, runtime
Model=HIRLAM20, 2006-05-24 00:00:00

#-----------------
# New position
#-----------------
#- Name of position, longitude, latitude
Position=GARDERMOEN, 11.1, 60.2

#- Data-header
#  the Time and Level specification is mandatory
#  Add your own parameters after these (defined properly in etc/parameters.def)
Time                  Level     MSLP    TT
#- Data-table matching header
2006-05-23 18:00:00   0         1000.1  23.3
2006-05-24 00:00:00   0         999.9   16.4
.....

#-----------------
# Start a new position with Position=
#-----------------
#- Name of position, longitude, latitude
Position=X,....
....
# End of Ascii file format description




debian/changelog
tseries (5.1.2-ubuntu2) precise; urgency=low

  * first version adapted to precise 64 bit 

 -- Juergen Schulze <j.schulze@met.no>  Tue, 07 Aug 2012 12:52:19 +0200

tseries (5.1.2-ubuntu1) lucid; urgency=low

  * fixed invisible axis
  * fixed vanishing stations

 -- Juergen Schulze <j.schulze@met.no>  Wed, 09 Nov 2011 12:04:24 +0100

tseries (5.1.1-ubuntu1) lucid; urgency=low

  * fixed sidebar 

 -- Juergen Schulze <j.schulze@met.no>  Mon, 24 Oct 2011 08:39:48 +0200

tseries (5.1.0-ubuntu3) lucid; urgency=low

  * get observations in utc - not nmt

 -- Juergen Schulze <j.schulze@met.no>  Fri, 14 Oct 2011 10:45:02 +0200

tseries (5.1.0-ubuntu2) lucid; urgency=low

  * setup fixes - wind in knots 

 -- Juergen Schulze <j.schulze@met.no>  Fri, 23 Sep 2011 08:25:33 +0200

tseries (5.1.0-ubuntu1) lucid; urgency=low

  * Interface to read from klimadb implemented
  * new style for EPS precipitation plots

 -- Juergen Schulze <j.schulze@met.no>  Mon, 19 Sep 2011 09:07:29 +0200

tseries (5.0.1-ubuntu4) lucid; urgency=low

  * fixed deafukt printer problem 

 -- Juergen Schulze <j.schulze@met.no>  Thu, 23 Sep 2010 11:27:04 +0200

tseries (5.0.1-ubuntu3) lucid; urgency=low

  *  more setup trouble under debian 

 -- Juergen Schulze <j.schulze@met.no>  Mon, 20 Sep 2010 12:54:12 +0200

tseries (5.0.1-ubuntu2) lucid; urgency=low

  * new ubuntu release for lucid

 -- Juergen Schulze <j.schulze@met.no>  Mon, 20 Sep 2010 12:16:10 +0200

tseries (5.0.1-ubuntu1) hardy; urgency=low

  * wdb support debugged and ready to release

 -- Juergen Schulze <j.schulze@met.no>  Mon, 31 May 2010 10:40:09 +0200

tseries (5.0.0-ubuntu1) hardy; urgency=low

  * tseries with on demand interpolation by WDB support

 -- Juergen Schulze <j.schulze@met.no>  Mon, 10 May 2010 08:52:04 +0200

tseries (4.1.0-ubuntu2) hardy; urgency=low

  * Full Qt4 transformation - disabled Qt3 support

 -- Juergen Schulze <j.schulze@met.no>  Mon, 22 Feb 2010 13:43:04 +0100

tseries (4.1.0-ubuntu1) stable; urgency=low

  * Initial release
  * tseries-test_4.1.0-ubuntu1_i386.deb

 -- Juergen Schulze <juergens@tko2698.pc.met.no>  Sun, 29 Nov 2009 20:30:36 +0100


src/tseries.cc
/*
  Tseries - A Free Meteorological Timeseries Viewer

  $Id$

  Copyright (C) 2006 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of Tseries

  Tseries is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  Tseries is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Tseries; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include "tsSetup.h"
#include "tsConfigure.h"
#include "qtsMain.h"

#include <fstream>

#include <puTools/miCommandLine.h>
#include <puTools/miTime.h>
#include <puTools/miString.h>



#include <QTranslator>
#include <QApplication>

using namespace std;
using namespace miutil;

int main(int argc, char **argv)
{
  vector<miCommandLine::option> o;
  o.push_back(miCommandLine::option( 's',"setup"    , 1 ));
  o.push_back(miCommandLine::option( 'h',"help"     , 0 ));
  o.push_back(miCommandLine::option( 'f',"file"     , 1 ));
  o.push_back(miCommandLine::option( 'S',"site"     , 1 ));
  o.push_back(miCommandLine::option( 'l',"lang"     , 1 ));
  o.push_back(miCommandLine::option( 'T',"title"    , 1 ));
  o.push_back(miCommandLine::option( 'd',"define"   , 1 ));
  o.push_back(miCommandLine::option( 'H',"host"     , 1 ));
  o.push_back(miCommandLine::option( 'u',"user"     , 1 ));

  miCommandLine cl(o,argc, argv);

   // read setup

  miString    site      = "FOU";
  miString    setupfile = "tseries.ctl";
  miString    title     = "T-series";
  tsSetup     setup;
  tsConfigure config;
  miString    lang;

  if(cl.hasFlag('h')){
    cerr << "Usage: "
         << argv[0]
         << "  -s setupfile "  << endl
         << "  -S site      "  << endl
         << "  -l lang      "  << endl
         << "  -T title     "  << endl
         << "  -H wdbhost   "  << endl
         << "  -u wdbuser   "  << endl
         << "  -d section1:key1=token1 section2:key2=token2 "
         << endl << endl;
    exit (0);
  }

  if(cl.hasFlag('S')) site         = cl.arg('S')[0];
  if(cl.hasFlag('s')) setupfile    = cl.arg('s')[0];
  if(cl.hasFlag('T')) title += " " + cl.arg('T')[0];


  if(!setup.read(setupfile,site))
    exit(0);

  if(setup.lang.exists())
    lang=setup.lang;

  config.read(setup.files.configure);
  config.get("LANG",lang);

  if(cl.hasFlag('d')) {
    vector<miString> overridetokens=cl.arg('d');
    for(int i=0; i<overridetokens.size();i++)
      setup.overrideToken(overridetokens[i]);
  }

  if(cl.hasFlag('H')) setup.overrideToken("WDB:host="+cl.arg('H')[0]);
  if(cl.hasFlag('u')) setup.overrideToken("WDB:user="+cl.arg('u')[0]);


  if(cl.hasFlag('l')) lang=cl.arg('l')[0];


  QApplication a( argc, argv );
  QTranslator  myapp( 0 );
  QTranslator  qt( 0 );

  miTime defTime;
  defTime.setDefaultLanguage(lang);

  if(setup.gui.style.exists())
    a.setStyle(setup.gui.style.cStr());


  if(lang.exists()) {

    miString qtlang   = "qt_"     +lang;
    miString langfile = "tseries_"+lang;

    // translation file for application strings

    for(unsigned int i=0;i<setup.path.lang.size(); i++ )
      if( qt.load( qtlang.cStr(),setup.path.lang[i].cStr()))
	break;

    for(unsigned int i=0;i<setup.path.lang.size(); i++ )
      if( myapp.load( langfile.cStr(),setup.path.lang[i].cStr()))
	break;

    a.installTranslator( &qt    );
    a.installTranslator( &myapp );

  }


  qtsMain *main = new qtsMain(lang);

  main->setWindowTitle(title.cStr());
  main->show();

  return a.exec();
}
